src/gui/TextEditor.cpp

```diff
@@ -904,8 +905,66 @@ auto TextEditor::blinkCallback(TextEditor* te) -> gint {
     return false;
 }
 
+void TextEditor::setTextToPangoLayout(PangoLayout* pl) const {
+    if (!this->preeditString.empty()) {
+        GtkTextIter cursorIter = {nullptr};
+        GtkTextMark* cursor = gtk_text_buffer_get_insert(this->buffer);
+        gtk_text_buffer_get_iter_at_mark(this->buffer, &cursorIter, cursor);
+
+        string text = this->text->getText();
+        int offset = gtk_text_iter_get_offset(&cursorIter);
+        int pos = gtk_text_iter_get_line_index(&cursorIter);
+
+        for (gtk_text_iter_set_line_index(&cursorIter, 0); gtk_text_iter_backward_line(&cursorIter);) {
+            pos += gtk_text_iter_get_bytes_in_line(&cursorIter);
+        }
+        gtk_text_iter_set_offset(&cursorIter, offset);
+        string txt = text.substr(0, pos) + preeditString + text.substr(pos);
+
+        PangoAttrList* attrlist = pango_attr_list_new();
+        PangoAttrList* preedit_attrlist = this->preeditAttrList;
+        pango_attr_list_splice(attrlist, preedit_attrlist, pos, preeditString.length());
+        pango_layout_set_attributes(pl, attrlist);
+        pango_attr_list_unref(attrlist);
+        attrlist = nullptr;
+        pango_layout_set_text(pl, txt.c_str(), txt.length());
+    } else {
+        string txt = this->text->getText();
+        pango_layout_set_text(pl, txt.c_str(), txt.length());
+    }
+}
+
+auto TextEditor::computeBoundingRect() -> Rectangle<double> {
+    /**
+     * We draw on a fake surface to get the size of the printed text
+     * See also TextView::calcSize
+     *
+     * NB: we cannot rely on TextView::calcSize directly, since it would not take the IM preeditString into account.
+     */
+    cairo_surface_t* surface = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, 1, 1);
+    cairo_t* cr = cairo_create(surface);
+    auto* textElement = this->getText();
+
+    PangoLayout* pl = TextView::initPango(cr, textElement);
+
+    setTextToPangoLayout(pl);
+
+    int w = 0;
+    int h = 0;
+    pango_layout_get_size(pl, &w, &h);
+    double width = (static_cast<double>(w)) / PANGO_SCALE;
+    double height = (static_cast<double>(h)) / PANGO_SCALE;
+    g_object_unref(pl);
+
+    cairo_destroy(cr);
+    cairo_surface_destroy(surface);
+
+    return Rectangle<double>(textElement->getX(), textElement->getY(), width, height);
+}
+
+
 void TextEditor::repaintEditor() {
-    auto rect = this->text->boundingRect();
+    auto rect = this->computeBoundingRect();
```

Comment:
In the future, we may want to consider caching the bounding rect, since it will currently be recalculated even if the contents do not change (e.g., cursor blink or moving the cursor will cause the bounding rect to be recalculated). Not obvious what type of performance impact this will cause--but I think we are currently still much better compared to when we were rerendering the whole page.

Reply:
I agree. I think there are a lot of possible optimizations of this class. Overall, a global refactor would be useful (e.g. removing GtkTextBuffer and use std::string instead -- or maybe a stream, using Pango directly for cursor position, and more)